File: snippets/fsharp/System/Type/GetMethod/GetMethodWithOverloads2.fs
module GetMethodWithOverloads2

// <Snippet3>
open System
open System.Reflection

type Person() =
    member val FirstName = "" with get, set
    member val LastName = "" with get, set

    override this.ToString() =
        (this.FirstName + " " + this.LastName).Trim()

let retrieveMethod (t: Type) name =
    try
        let m = t.GetMethod name
        if m <> null then
            printfn $"""{m.ReflectedType.Name}.{m.Name}: {if m.IsStatic then "Static" else "Instance"} method"""
        else
            printfn $"{t.Name}.ToString method not found"
    with :? AmbiguousMatchException ->
        printfn $"{t.Name}.{name} has multiple public overloads."

let t = typeof<Person>
retrieveMethod t "ToString"

let t2 = typeof<int>
retrieveMethod t2 "ToString"

// The example displays the following output:
//       Person.ToString: Instance method
//       Int32.ToString has multiple public overloads.
// </Snippet3>

File: snippets/csharp/System/Type/GetMethod/GetMethodWithOverloads2.cs
// <Snippet3>
using System;
using System.Reflection;

public class Person
{
    public string FirstName;
    public string LastName;

    public override string ToString() => (FirstName + " " + LastName).Trim();
}

public class Example2
{
    public static void Main()
    {
        Type t = typeof(Person);
        RetrieveMethod(t, "ToString");

        t = typeof(int);
        RetrieveMethod(t, "ToString");
    }

    private static void RetrieveMethod(Type t, string name)
    {
        try
        {
            MethodInfo m = t.GetMethod(name);
            if (m != null)
                Console.WriteLine($"{m.ReflectedType.Name}.{m.Name}: {(m.IsStatic ? "Static" : "Instance")} method");
            else
                Console.WriteLine($"{t.Name}.ToString method not found");
        }
        catch (AmbiguousMatchException)
        {
            Console.WriteLine($"{t.Name}.{name} has multiple public overloads.");
        }
    }
}
// The example displays the following output:
//       Person.ToString: Instance method
//       Int32.ToString has multiple public overloads.
// </Snippet3>
